algosdk/abi.py

```diff
@@ -0,0 +1,570 @@
+            ufixed_regex = "^ufixed([1-9][\d]*)x([1-9][\d]*)$"
+            matches = re.search(ufixed_regex, s)
+            try:
+                bit_size = int(matches.group(1))
+                precision = int(matches.group(2))
+                return Type.make_ufixed_type(bit_size, precision)
+            except:
+                raise error.ABITypeError(f"malformed ufixed string: {s}")
+        elif s == "bool":
+            return Type.make_bool_type()
+        elif s == "address":
+            return Type.make_address_type()
+        elif s == "string":
+            return Type.make_string_type()
+        elif len(s) >= 2 and s[0] == "(" and s[-1] == ")":
+            # Recursively parse parentheses from a tuple string
+            tuples = Type.parse_tuple(s[1:-1])
+            tuple_list = []
+            for tup in tuples:
+                if isinstance(tup, str):
+                    tt = Type.type_from_string(str(tup))
+                    tuple_list.append(tt)
+                elif isinstance(tup, list):
+                    tts = list(map(lambda t_: Type.type_from_string(t_), tup))
+                    tuple_list.append(tts)
+                else:
+                    raise error.ABITypeError(
+                        f"cannot convert {tup} to an ABI type"
+                    )
+
+            return Type.make_tuple_type(tuple_list)
+        else:
+            raise error.ABITypeError(f"cannot convert {s} to an ABI type")
+
+    def parse_tuple(s):
+        """
+        Given a tuple string, parses one layer of the tuple.
+        i.e. '(x,(y,(z)))' -> ['x', '(y,(z))']
+        """
+        # If the tuple content is empty, return an empty list
+        if not s:
+            return []
+
+        if s.startswith(",") or s.endswith(","):
+            raise error.ABITypeError(
+                f"cannot have leading or trailing commas in {s}"
+            )
+
+        if ",," in s:
+            raise error.ABITypeError(f"cannot have consecutive commas in {s}")
+
+        tuple_strs = []
+        depth = 0
+        word = ""
+        for char in s:
+            word += char
+            if char == "(":
+                depth += 1
+            elif char == ")":
+                depth -= 1
+            elif char == ",":
+                # If the comma is at depth 0, put the word as a new token.
+                if depth == 0:
+                    word = word[:-1]
+                    if word:
+                        tuple_strs.append(word)
+                        word = ""
+        if word:
+            tuple_strs.append(word)
+        if depth != 0:
+            raise error.ABITypeError(f"parenthesis mismatch: {s}")
+        return tuple_strs
+
+    def is_dynamic(self):
+        """
+        Returns whether the current type or any of its child types is a
+        dynamic type or not.
+        """
+        if (
+            self.abi_type_id == BaseType.ArrayDynamic
+            or self.abi_type_id == BaseType.String
+        ):
```

Comment:
super nit: `self.abi_type_id in (BaseType.ArrayDynamic, BaseType.String)` is more concise.

Similarly, for below: `return any(child.is_dynamic() for child in self.child_types)`